teps/0041-tekton-component-versioning.md

```diff
@@ -0,0 +1,304 @@
+                  type: string
+                  description: version of the component
+  scope: Cluster
+  names:
+    plural: info
+    singular: info
+    kind: Info
+    categories:
+      - tekton
+      - tekton-pipelines
+```
+
+And a CR should be added during installation which has the details
+about the version of the component or any other info required.
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Info
+metadata:
+  name: pipeline-info
+spec:
+  version: v0.18.0
+```
+
+`ClusterRole` and `ClusterRoleBinding`
+
+```yaml
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: pipelines-info-view-cr
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+rules:
+  - apiGroups: ["tekton.dev"]
+    resources: ["info"]
+    verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: pipelines-info-crbinding
+subjects:
+  - kind: Group
+    name: system:authenticated
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: pipelines-info-view
+```
+
+#### Advantages
+
+1. The CRD can be clusterwide so user need not to know the installed pipeline namespace.
+
+#### Disadvantages
+
+1. Where to store the CRD?
+
+   - Pipelines, Triggers or somewhere else?
+   - In each component then how to handle the changes?
+     - (Solution) Each component will have to maintain their own CRD so that they
+       can provide their respective information if they want to.
+
+### 2. Creating a ConfigMap
+
+Tekton components should create their own ConfigMap which contains
+the version information.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: pipeline-version
+  namespace: tekton-pipelines
+data:
+  version: v0.19.0
+```
+
+`Role` and `RoleBinding`
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: info
+  namespace: tekton-pipelines
+rules:
+  - apiGroups: [""]
+    resources: ["configmaps"]
+    resourceNames: ["pipelines-version"]
+    verbs: ["get", "describe"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: info
+  namespace: tekton-pipelines
+subjects:
+  - kind: Group
+    name: system:authenticated
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: info
+```
+
+#### Advantages
+
+1. Other users not having access to the namespace can still have access
+   to the particular ConfigMap.
+
+   `kubectl get cm pipelines-version -n tekton-pipelines`
+
+#### Disadvantages
+
+1. User need to know the namespace in which Tekton components are installed
+   since the scope is only the namespace.
+
+## Test Plan
+
+Basic test scenarios of getting the version information should be added in all the components.
+Other components like CLI can add the test cases based on their usage.
+
+## Design Evaluation
+
+## Drawbacks
+
+## Alternatives
```

Comment:
possible alternatives
* provide an endpoint from the controller that can be queried and will return what version it is (maybe via mechanism similar to aliveness probes?) <-- if this is something we could actually do, it would guarantee that the version matches the installed version (vs having 2 separate pieces to keep in sync)
* continue to rely on querying for labels (looks like that's the approach knative has taken? https://knative.dev/docs/install/check-install-version/)
* Could we provide a role & rolebinding as part of installing tekton that allows for read access to the existing version labels? if not, can the problem statement go into more detail about why this is not reasonable? it seems like we need to provide role and rolebinding regardless

Reply:
> possible alternatives
> 
> * provide an endpoint from the controller that can be queried and will return what version it is (maybe via mechanism similar to aliveness probes?) <-- if this is something we could actually do, it would guarantee that the version matches the installed version (vs having 2 separate pieces to keep in sync)

+1

> * continue to rely on querying for labels (looks like that's the approach knative has taken? https://knative.dev/docs/install/check-install-version/)
> * Could we provide a role & rolebinding as part of installing tekton that allows for read access to the existing version labels? if not, can the problem statement go into more detail about why this is not reasonable? it seems like we need to provide role and rolebinding regardless

Reply:
> Could we provide a role & rolebinding as part of installing tekton that allows for read access to the existing version labels? if not, can the problem statement go into more detail about why this is not reasonable? it seems like we need to provide role and rolebinding regardless

I think this would require giving read access to the entire deployment?